lib/chef/resource/systemd_service.rb

```diff
@@ -0,0 +1,49 @@
+#
+# Author:: Davide Cavalca (<dcavalca@fb.com>)
+# Copyright:: Copyright (c) 2015 Facebook
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/resource/service'
+require 'chef/provider/service/systemd'
+
+class Chef
+  class Resource
+    class SystemdService < Chef::Resource::Service
+      provides :service, os: "linux" do |node|
+        Chef::Platform::ServiceHelpers.service_resource_providers.include?(:systemd)
+      end
```

Comment:
This is probably not the right approach to take for the service resource.  We lazy the evaluation of what kind of provider to invoke to converge time so that we can inspect the service and determine the correct provider.  With this change you may wind up with systemd resources wrapping non-systemd providers.

Another way of looking at this is that instead of doing provider resolver stuff with mutiple object classes for the service provider, what we maybe should have done is collapse the service provider down to a single resource which at converge time would choose the correct implementation to run based on inspecting the system.  That would keep the model 1:1 but delay the decision of the implementation to converge time.

In either case you lose the ability to do systemd-specific validation on the resource, early.  But you can't ever really do that because you don't know if you've really got a systemd service or not until converge time, long after the resource has been constructed.

So this is trying to do the right thing to get validation correct in the resource -- and normally i would approve of moving to a more 1:1 mapping between resource and provider -- but this is the one case that breaks that model.